flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java

```diff
@@ -0,0 +1,359 @@
+			}
+		} else {
+			GroupType parquetGroupType = fieldType.asGroupType();
+			OriginalType originalType = parquetGroupType.getOriginalType();
+			if (originalType != null) {
+				switch (originalType) {
+					case LIST:
+						if (parquetGroupType.getFieldCount() != 1) {
+							throw new UnsupportedOperationException("Invalid list type " + parquetGroupType);
+						}
+						Type repeatedType = parquetGroupType.getType(0);
+						if (!repeatedType.isRepetition(Type.Repetition.REPEATED)) {
+							throw new UnsupportedOperationException("Invalid list type " + parquetGroupType);
+						}
+
+						if (repeatedType.isPrimitive()) {
+							// Backward-compatibility element group doesn't exist also allowed
+							typeInfo = BasicArrayTypeInfo.getInfoFor(
+								Array.newInstance(convertField(repeatedType).getTypeClass(), 0).getClass());
+						} else {
+							// Backward-compatibility element group name can be any string (element/array/other)
+							GroupType elementType = repeatedType.asGroupType();
+							if (elementType.getFieldCount() > 1) {
+								typeInfo = ObjectArrayTypeInfo.getInfoFor(convertField(elementType));
+							} else {
+								Type internalType = elementType.getType(0);
+								if (internalType.isPrimitive()) {
+									typeInfo = BasicArrayTypeInfo.getInfoFor(
+										Array.newInstance(convertField(internalType).getTypeClass(),
+											0).getClass());
+								} else {
+									typeInfo = ObjectArrayTypeInfo.getInfoFor(convertField(internalType));
+								}
+							}
+						}
+						break;
+
+					case MAP_KEY_VALUE:
+					case MAP:
+						if (parquetGroupType.getFieldCount() != 1 || parquetGroupType.getType(0).isPrimitive()) {
+							throw new UnsupportedOperationException("Invalid map type " + parquetGroupType);
+						}
+
+						GroupType mapKeyValType = parquetGroupType.getType(0).asGroupType();
+						if (!mapKeyValType.isRepetition(Type.Repetition.REPEATED)
+							|| mapKeyValType.getFieldCount() != 2) {
+							throw new UnsupportedOperationException("Invalid map type " + parquetGroupType);
+						}
+						Type keyType = mapKeyValType.getType(0);
+						if (!keyType.isPrimitive()
+							|| !keyType.asPrimitiveType().getPrimitiveTypeName().equals(
+								PrimitiveType.PrimitiveTypeName.BINARY)
+							|| !keyType.getOriginalType().equals(OriginalType.UTF8)) {
+							throw new IllegalArgumentException("Map key type must be binary (UTF8): "
+								+ keyType);
+						}
+
+						Type valueType = mapKeyValType.getType(1);
+						return new MapTypeInfo<>(BasicTypeInfo.STRING_TYPE_INFO, convertField(valueType));
+					default:
+						throw new UnsupportedOperationException("Unsupported schema: " + fieldType);
+				}
+			} else {
+				// if no original type than it is a record
+				return convertFields(parquetGroupType.getFields());
+			}
+		}
+
+		return typeInfo;
+	}
+
+	private static Type convertField(String fieldName, TypeInformation<?> typeInfo,
```

Comment:
I assume only the physical type is important here?
Or do we need to be precise and provide the logical type as well because the schema is stored in the file?

Reply:
As suggested by Timo, this function considers both physical type and logical type. As you can see, it firstly convert type by physical type, if there are multiple logical types using the same physical data size to store (for example TIME_MICROS, TIMESTAMP_MICROS are used as INT64).